src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1842,26 +1843,17 @@ SingleShardSelectTaskList(Query *query, uint64 jobId, List *relationShardList,
 						  List *placementList, uint64 shardId, bool localFastPathQuery)
 {
 	Task *task = CreateTask(SELECT_TASK);
-	StringInfo queryString = makeStringInfo();
 	List *relationRowLockList = NIL;
 
 	RowLocksOnRelations((Node *) query, &relationRowLockList);
 
 	/*
-	 * For performance reasons, we skip generating the queryString for local
-	 * fast path queries. With this, we can avoid deparsing during the execution
-	 * as well.
+	 * For performance reasons, we skip generating the queryString. For local
+	 * execution this is not needed, so we wait until the executor determines
+	 * that the query cannot be executed locally.
 	 */
-	if (localFastPathQuery)
-	{
-		queryString = NULL;
-	}
-	else
-	{
-		pg_get_query_def(query, queryString);
-	}
-
-	task->queryString = queryString ? queryString->data : NULL;
+	task->query = copyObject(query);
```

Comment:
Copying the query tree seems like an expensive operation. Can it be avoided? The Job also has a query tree.

Reply:
We can at least use the job query tree in some cases. And possibly we don't need to copy the query if we already copy it later on. Since we don't actually modify the query in the task. I'm mainly copying it currently to rule it out as a source for the errors.

Reply:
I've stopped copying the querytree and all the tests keep passing, so I think it's fine to not copy them. I'll investigate a bit more though to be more sure.